File: android/packages/components/src/main/java/io/ente/components/Actions.kt
package io.ente.components

import androidx.compose.animation.core.animateFloatAsState
import androidx.compose.animation.core.tween
import androidx.compose.foundation.clickable
import androidx.compose.foundation.interaction.MutableInteractionSource
import androidx.compose.foundation.interaction.collectIsPressedAsState
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.heightIn
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.luminance
import androidx.compose.ui.graphics.graphicsLayer
import androidx.compose.ui.semantics.Role
import androidx.compose.ui.text.style.TextDecoration
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp

enum class ActionKind { Primary, Secondary, Neutral, Destructive, DestructiveText, Link }

enum class ActionSize { Regular, Compact }

enum class IconActionKind { Primary, Critical, Unfilled, Secondary, Accent, Circular }

enum class FloatingActionKind { Primary, Secondary, Destructive }

@Composable
fun ActionButton(
    label: String,
    onClick: () -> Unit,
    modifier: Modifier = Modifier,
    kind: ActionKind = ActionKind.Primary,
    size: ActionSize = ActionSize.Regular,
    enabled: Boolean = true,
    fullWidth: Boolean = true,
    leading: @Composable (() -> Unit)? = null,
) {
    val palette = LocalEntePalette.current
    val interactionSource = remember { MutableInteractionSource() }
    val pressed by interactionSource.collectIsPressedAsState()
    val colors = actionColors(kind, palette, enabled, pressed)
    val scale by animateFloatAsState(
        targetValue = if (pressed && enabled) 0.98f else 1f,
        animationSpec = tween(EnteMotion.quick),
        label = "actionScale",
    )
    val verticalPadding = if (size == ActionSize.Regular) 14.dp else 12.dp
    val actionModifier = if (fullWidth) modifier.fillMaxWidth() else modifier

    Surface(
        color = colors.background,
        contentColor = colors.foreground,
        shape = androidx.compose.foundation.shape.RoundedCornerShape(EnteRadius.button),
        modifier = actionModifier
            .graphicsLayer { scaleX = scale; scaleY = scale }
            .clickable(
                enabled = enabled,
                role = Role.Button,
                interactionSource = interactionSource,
                indication = null,
                onClick = onClick,
            ),
    ) {
        Row(
            modifier = Modifier
                .heightIn(min = 24.dp)
                .padding(horizontal = EnteSpacing.xl, vertical = verticalPadding),
            horizontalArrangement = Arrangement.Center,
            verticalAlignment = Alignment.CenterVertically,
        ) {
            if (leading != null) {
                Box(Modifier.size(18.dp), contentAlignment = Alignment.Center) { leading() }
                androidx.compose.foundation.layout.Spacer(Modifier.width(EnteSpacing.sm))
            }
            Text(
                text = label,
                maxLines = 2,
                overflow = TextOverflow.Ellipsis,
                style = if (size == ActionSize.Regular) EnteTypography.bodyBold else EnteTypography.body,
                textDecoration = if (kind == ActionKind.DestructiveText || kind == ActionKind.Link) TextDecoration.Underline else null,
            )
        }
    }
}

@Composable
fun FloatingAction(
    onClick: () -> Unit,
    modifier: Modifier = Modifier,
    kind: FloatingActionKind = FloatingActionKind.Primary,
    content: @Composable () -> Unit,
) {
    val palette = LocalEntePalette.current
    val interactionSource = remember { MutableInteractionSource() }
    val pressed by interactionSource.collectIsPressedAsState()
    val scale by animateFloatAsState(
        targetValue = if (pressed) 0.98f else 1f,
        animationSpec = tween(EnteMotion.quick),
        label = "floatingActionScale",
    )
    val colors = floatingActionColors(kind, palette, pressed)
    Surface(
        color = colors.background,
        contentColor = colors.foreground,
        shape = androidx.compose.foundation.shape.CircleShape,
        modifier = modifier
            .size(52.dp)
            .graphicsLayer { scaleX = scale; scaleY = scale }
            .clickable(
                role = Role.Button,
                interactionSource = interactionSource,
                indication = null,
                onClick = onClick,
            ),
    ) {
        Box(contentAlignment = Alignment.Center) { content() }
    }
}

@Composable
fun IconAction(
    onClick: () -> Unit,
    modifier: Modifier = Modifier,
    kind: IconActionKind = IconActionKind.Secondary,
    enabled: Boolean = true,
    size: Dp = 40.dp,
    content: @Composable () -> Unit,
) {
    val palette = LocalEntePalette.current
    val interactionSource = remember { MutableInteractionSource() }
    val pressed by interactionSource.collectIsPressedAsState()
    val colors = actionColors(kind, palette, enabled, pressed)
    val scale by animateFloatAsState(
        targetValue = if (pressed && enabled) 0.98f else 1f,
        animationSpec = tween(EnteMotion.quick),
        label = "iconActionScale",
    )

    Surface(
        color = colors.background,
        contentColor = colors.foreground,
        shape = androidx.compose.foundation.shape.RoundedCornerShape(if (kind == IconActionKind.Circular) size / 2 else EnteRadius.medium),
        modifier = modifier
            .size(size)
            .graphicsLayer { scaleX = scale; scaleY = scale }
            .clickable(
                enabled = enabled,
                role = Role.Button,
                interactionSource = interactionSource,
                indication = null,
                onClick = onClick,
            ),
    ) {
        Box(contentAlignment = Alignment.Center) { content() }
    }
}

private data class ActionColors(val background: Color, val foreground: Color)

private fun actionColors(kind: ActionKind, palette: Palette, enabled: Boolean, pressed: Boolean): ActionColors {
    if (!enabled) {
        return ActionColors(
            if (kind == ActionKind.DestructiveText || kind == ActionKind.Link) Color.Transparent else palette.fill,
            palette.disabledText,
        )
    }
    return when (kind) {
        ActionKind.Primary -> ActionColors(if (pressed) palette.primaryDarker else palette.primary, Color.White)
        ActionKind.Secondary -> ActionColors(if (pressed) palette.fillDarkest else palette.fill, palette.text)
        ActionKind.Neutral -> ActionColors(palette.text, palette.reverseText)
        ActionKind.Destructive -> ActionColors(if (pressed) palette.dangerDarker else palette.danger, Color.White)
        ActionKind.DestructiveText -> ActionColors(Color.Transparent, if (pressed) palette.dangerDarker else palette.danger)
        ActionKind.Link -> ActionColors(Color.Transparent, if (pressed) palette.primaryDarker else palette.primary)
    }
}

private fun actionColors(kind: IconActionKind, palette: Palette, enabled: Boolean, pressed: Boolean): ActionColors {
    if (!enabled) {
        return ActionColors(
            if (kind == IconActionKind.Unfilled || kind == IconActionKind.Secondary) Color.Transparent else palette.fill,
            palette.hintText,
        )
    }
    return when (kind) {
        IconActionKind.Primary, IconActionKind.Circular -> ActionColors(if (pressed) palette.fillDarker else palette.surface, palette.text)
        IconActionKind.Critical -> ActionColors(if (pressed) palette.fillDarkest else palette.fill, palette.text)
        IconActionKind.Unfilled, IconActionKind.Secondary -> ActionColors(
            Color.Transparent,
            if (palette.background.luminance() < 0.5f) palette.text else palette.text.copy(alpha = 0.75f),
        )
        IconActionKind.Accent -> ActionColors(if (pressed) palette.primaryDarker else palette.primary, Color.White)
    }
}

private fun floatingActionColors(kind: FloatingActionKind, palette: Palette, pressed: Boolean): ActionColors = when (kind) {
    FloatingActionKind.Primary -> ActionColors(if (pressed) palette.primaryDark else palette.primary, Color.White)
    FloatingActionKind.Secondary -> ActionColors(if (pressed) palette.fillDarker else palette.fill, palette.primary)
    FloatingActionKind.Destructive -> ActionColors(if (pressed) palette.dangerDark else palette.danger, Color.White)
}

File: apple/packages/EnteComponents/Sources/EnteComponents/Actions.swift
import SwiftUI

public enum ActionKind {
    case primary
    case secondary
    case neutral
    case destructive
    case destructiveText
    case link
}

public enum ActionSize {
    case regular
    case compact
}

public enum IconActionKind {
    case primary
    case critical
    case unfilled
    case secondary
    case accent
    case circular
}

public enum FloatingActionKind {
    case primary
    case secondary
    case destructive
}

public struct ActionButton: View {
    @Environment(\.entePalette) private var palette
    private let title: String
    private let systemImage: String?
    private let kind: ActionKind
    private let size: ActionSize
    private let enabled: Bool
    private let fullWidth: Bool
    private let action: () -> Void

    public init(
        _ title: String,
        systemImage: String? = nil,
        kind: ActionKind = .primary,
        size: ActionSize = .regular,
        enabled: Bool = true,
        fullWidth: Bool = true,
        action: @escaping () -> Void
    ) {
        self.title = title
        self.systemImage = systemImage
        self.kind = kind
        self.size = size
        self.enabled = enabled
        self.fullWidth = fullWidth
        self.action = action
    }

    public var body: some View {
        Button(action: action) {
            Group {
                if let systemImage {
                    Label(title, systemImage: systemImage)
                } else {
                    Text(title)
                }
            }
                .lineLimit(2)
                .font(size == .regular ? EnteTypography.bodyBold : EnteTypography.body)
                .underline(kind == .destructiveText || kind == .link)
                .frame(maxWidth: fullWidth ? .infinity : nil, minHeight: 24)
                .padding(.horizontal, EnteSpacing.xl)
                .padding(.vertical, size == .regular ? 14 : 12)
                .frame(minHeight: 52)
        }
        .buttonStyle(ActionStyle(kind: kind, palette: palette))
        .disabled(!enabled)
    }
}

public struct IconAction<Content: View>: View {
    @Environment(\.entePalette) private var palette
    private let kind: IconActionKind
    private let enabled: Bool
    private let size: CGFloat
    private let action: () -> Void
    private let content: Content

    public init(
        kind: IconActionKind = .secondary,
        enabled: Bool = true,
        size: CGFloat = 40,
        action: @escaping () -> Void,
        @ViewBuilder content: () -> Content
    ) {
        self.kind = kind
        self.enabled = enabled
        self.size = size
        self.action = action
        self.content = content()
    }

    public var body: some View {
        Button(action: action) {
            content
                .frame(width: size, height: size)
        }
        .buttonStyle(IconActionStyle(kind: kind, palette: palette, cornerRadius: kind == .circular ? size / 2 : EnteRadius.medium))
        .disabled(!enabled)
    }
}

public struct FloatingAction<Content: View>: View {
    @Environment(\.entePalette) private var palette
    private let kind: FloatingActionKind
    private let action: () -> Void
    private let content: Content

    public init(
        kind: FloatingActionKind = .primary,
        action: @escaping () -> Void,
        @ViewBuilder content: () -> Content
    ) {
        self.kind = kind
        self.action = action
        self.content = content()
    }

    public var body: some View {
        Button(action: action) {
            content.frame(width: 52, height: 52)
        }
        .buttonStyle(FloatingActionStyle(kind: kind, palette: palette))
    }
}

private struct ActionStyle: ButtonStyle {
    @Environment(\.isEnabled) private var enabled
    let kind: ActionKind
    let palette: Palette

    func makeBody(configuration: Configuration) -> some View {
        configuration.label
            .foregroundStyle(foreground(configuration.isPressed))
            .background(background(configuration.isPressed))
            .clipShape(RoundedRectangle(cornerRadius: EnteRadius.button, style: .continuous))
            .scaleEffect(configuration.isPressed ? 0.98 : 1)
            .animation(.easeOut(duration: EnteMotion.quick), value: configuration.isPressed)
    }

    private func background(_ pressed: Bool) -> Color {
        if !enabled {
            return kind == .destructiveText || kind == .link ? .clear : palette.fill
        }
        return switch kind {
        case .primary: pressed ? palette.primaryDarker : palette.primary
        case .secondary: pressed ? palette.fillDarkest : palette.fill
        case .neutral: palette.text
        case .destructive: pressed ? palette.dangerDarker : palette.danger
        case .destructiveText, .link: .clear
        }
    }

    private func foreground(_ pressed: Bool) -> Color {
        if !enabled { return palette.disabledText }
        return switch kind {
        case .primary, .destructive: .white
        case .secondary: palette.text
        case .neutral: palette.reverseText
        case .destructiveText: pressed ? palette.dangerDarker : palette.danger
        case .link: pressed ? palette.primaryDarker : palette.primary
        }
    }
}

private struct IconActionStyle: ButtonStyle {
    @Environment(\.isEnabled) private var enabled
    let kind: IconActionKind
    let palette: Palette
    let cornerRadius: CGFloat

    func makeBody(configuration: Configuration) -> some View {
        configuration.label
            .foregroundStyle(foreground)
            .background(background(configuration.isPressed))
            .clipShape(RoundedRectangle(cornerRadius: cornerRadius, style: .continuous))
            .scaleEffect(configuration.isPressed ? 0.98 : 1)
            .animation(.easeOut(duration: EnteMotion.quick), value: configuration.isPressed)
    }

    private func background(_ pressed: Bool) -> Color {
        if !enabled {
            return kind == .unfilled || kind == .secondary ? .clear : palette.fill
        }
        return switch kind {
        case .primary, .circular: pressed ? palette.fillDarker : palette.surface
        case .critical: pressed ? palette.fillDarkest : palette.fill
        case .unfilled, .secondary: .clear
        case .accent: pressed ? palette.primaryDarker : palette.primary
        }
    }

    private var foreground: Color {
        if !enabled { return palette.hintText }
        return switch kind {
        case .accent: .white
        case .unfilled, .secondary: palette.text.opacity(palette.isDark ? 1 : 0.75)
        case .primary, .critical, .circular: palette.text
        }
    }
}

private struct FloatingActionStyle: ButtonStyle {
    let kind: FloatingActionKind
    let palette: Palette

    func makeBody(configuration: Configuration) -> some View {
        configuration.label
            .foregroundStyle(foreground)
            .background(background(configuration.isPressed))
            .clipShape(Circle())
            .scaleEffect(configuration.isPressed ? 0.98 : 1)
            .animation(.easeOut(duration: EnteMotion.quick), value: configuration.isPressed)
    }

    private func background(_ pressed: Bool) -> Color {
        return switch kind {
        case .primary: pressed ? palette.primaryDark : palette.primary
        case .secondary: pressed ? palette.fillDarker : palette.fill
        case .destructive: pressed ? palette.dangerDark : palette.danger
        }
    }

    private var foreground: Color {
        return switch kind {
        case .secondary: palette.primary
        case .primary, .destructive: .white
        }
    }
}
